# Bacterial genome annotation tutorial
This is primarily intended for single genomes, not metagenomes, although most of these tools have options for inputting metagenome sequences.

Basic steps are:
- trim sequences
- genome assembly (for this I will be doing de novo assembly, but you could also use a reference guided assembly)
- gene prediction and annotation
- downstream analysis (pangenome, AMR, GWAS, phage)
  
For this tutorial I have downloaded a random selection of 10 sequencing runs from the NCBI SRA, but this will work with any raw sequences. 
I have included links for all the tools below, and hopefully a python environment you can download to work in which will provide access to some of the tools.
